arch/x86/pagetables.c

```diff
@@ -147,6 +147,23 @@ static mfn_t get_pgentry_mfn(mfn_t tab_mfn, pt_index_t index, unsigned long flag
         tab = init_map_mfn(mfn);
         memset(tab, 0, PAGE_SIZE);
     }
+    else {
+        /* Valid page table already exists, but its current protection bits may conflict
+         * with our. Our new flags will take precedence over previous one. */
+        unsigned long entry_new = *entry;
+        entry_new |= flags & _PAGE_USER;
+        entry_new |= flags & _PAGE_RW;
+        entry_new &= ~(flags & _PAGE_NX);
+        if (*entry != entry_new) {
+            char flags_str_old[16];
+            char flags_str_new[16];
+            printk("WARNING: Already-present page table protection conflicts with our.\n"
+                   "         Updating present page table protection: %s -> %s\n",
+                   dump_pte_flags(flags_str_old, 16, (pte_t) *entry),
+                   dump_pte_flags(flags_str_new, 16, (pte_t) entry_new));
+            *entry = entry_new;
```

Comment:
this might require a TLB flush (not to mention TLB shootdown, which is badly missing in general). Although the TLB will be flushed with the `set_pgentry()` for the lower level page table entry, it might break when we start using `invalpg`